File: src/main/java/de/hanno/hpengine/renderer/drawstrategy/FirstPassResult.java
package de.hanno.hpengine.renderer.drawstrategy;

import java.util.HashMap;
import java.util.Map;

public final class FirstPassResult {
    public int verticesDrawn;
    public int entitiesDrawn;
    public int linesDrawn;
    public boolean directionalLightShadowMapWasRendered;
    private Map<String, Object> properties = new HashMap<>();
    private boolean notYetUploadedVertexBufferDrawn;

    public FirstPassResult() {
    }

    public void init(int verticesDrawn, int entityCount, int linesDrawn, boolean directionalLightShadowMapWasRendered, boolean notYetUploadedVertexBufferDrawn) {
        this.verticesDrawn = verticesDrawn;
        this.entitiesDrawn = entityCount;
        this.linesDrawn = linesDrawn;
        this.directionalLightShadowMapWasRendered = directionalLightShadowMapWasRendered;
        this.notYetUploadedVertexBufferDrawn = notYetUploadedVertexBufferDrawn;
    }

    public void reset() {
        init(0,0,0,false, false);
    }

    public void setProperty(String vctLightInjectedFramesAgo, Object value) {
        properties.put(vctLightInjectedFramesAgo, value);
    }

    public Object getProperty(String key) {
        return properties.get(key);
    }

    public Map<String, Object> getProperties() {
        return properties;
    }

    public void set(FirstPassResult firstPassResult) {
        init(firstPassResult.verticesDrawn, firstPassResult.entitiesDrawn, firstPassResult.linesDrawn, firstPassResult.directionalLightShadowMapWasRendered, firstPassResult.notYetUploadedVertexBufferDrawn);
    }
}

File: engine/src/main/kotlin/de/hanno/hpengine/engine/graphics/renderer/drawstrategy/FirstPassResult.kt
package de.hanno.hpengine.engine.graphics.renderer.drawstrategy

class FirstPassResult {
    var verticesDrawn = 0
    var entitiesDrawn = 0
    var linesDrawn = 0
    var directionalLightShadowMapWasRendered = false
    private val properties: MutableMap<String, Any> = HashMap()
    private var notYetUploadedVertexBufferDrawn = false
    fun init(
        verticesDrawn: Int,
        entityCount: Int,
        linesDrawn: Int,
        directionalLightShadowMapWasRendered: Boolean,
        notYetUploadedVertexBufferDrawn: Boolean
    ) {
        this.verticesDrawn = verticesDrawn
        entitiesDrawn = entityCount
        this.linesDrawn = linesDrawn
        this.directionalLightShadowMapWasRendered = directionalLightShadowMapWasRendered
        this.notYetUploadedVertexBufferDrawn = notYetUploadedVertexBufferDrawn
    }

    fun reset() {
        init(0, 0, 0, false, false)
    }

    fun setProperty(vctLightInjectedFramesAgo: String, value: Any) {
        properties[vctLightInjectedFramesAgo] = value
    }

    fun getProperty(key: String): Any? {
        return properties[key]
    }

    fun getProperties(): Map<String, Any> {
        return properties
    }

    fun set(firstPassResult: FirstPassResult) {
        init(
            firstPassResult.verticesDrawn,
            firstPassResult.entitiesDrawn,
            firstPassResult.linesDrawn,
            firstPassResult.directionalLightShadowMapWasRendered,
            firstPassResult.notYetUploadedVertexBufferDrawn
        )
    }
}